## Record con CustomerID nulli 

In [30]:
import numpy as np
import pandas as pd

In [31]:
df = pd.read_csv("dataset/TC-dataset.csv", sep="\t",index_col=0)
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 471910 entries, 0 to 541909
Data columns (total 8 columns):
BasketID           471910 non-null object
BasketDate         471910 non-null object
Sale               471910 non-null float64
CustomerID         406830 non-null float64
CustomerCountry    471910 non-null object
ProdID             471910 non-null object
ProdDescr          471157 non-null object
Qta                471910 non-null int64
dtypes: float64(2), int64(1), object(5)
memory usage: 32.4+ MB


In [32]:
df[df["CustomerID"].isnull()]

BasketID           BasketDate     Sale  CustomerID CustomerCountry  \
1444     536544  2010-12-01 14:32:00     2.51         NaN  United Kingdom   
1446     536544  2010-12-01 14:32:00     1.66         NaN  United Kingdom   
1447     536544  2010-12-01 14:32:00     1.66         NaN  United Kingdom   
1448     536544  2010-12-01 14:32:00     2.51         NaN  United Kingdom   
1450     536544  2010-12-01 14:32:00     0.43         NaN  United Kingdom   
...         ...                  ...      ...         ...             ...   
541532   581498  2011-12-09 10:26:00     4.13         NaN  United Kingdom   
541534   581498  2011-12-09 10:26:00     3.29         NaN  United Kingdom   
541538   581498  2011-12-09 10:26:00     4.96         NaN  United Kingdom   
541539   581498  2011-12-09 10:26:00    10.79         NaN  United Kingdom   
541540   581498  2011-12-09 10:26:00  1714.17         NaN  United Kingdom   

        ProdID                           ProdDescr  Qta  
1444     21774     DECORATIVE CATS BATHROOM BOTTLE    2  
1446     21787               RAIN PONCHO RETROSPOT    2  
1447     21790                  VINTAGE SNAP CARDS    9  
1448     21791  VINTAGE HEADS AND TAILS CARD GAME     2  
1450     21802     CHRISTMAS TREE HEART DECORATION    9  
...        ...                                 ...  ...  
541532   85038     6 CHOCOLATE LOVE HEART T-LIGHTS    1  
541534  85049a       TRADITIONAL CHRISTMAS RIBBONS    5  
541538   85150       LADIES & GENTLEMEN METAL SIGN    1  
541539   85174                   S/4 CACTI CANDLES    1  
541540     DOT                      DOTCOM POSTAGE    1  

[65080 rows x 8 columns]

In [33]:
import random

basket_list_customer_null=df[df["CustomerID"].isnull()].BasketID.unique()
new_customer_per_basket_list=random.sample(range(1, 100000), len(basket_list_customer_null))


for i, elem in enumerate(basket_list_customer_null):
    df["CustomerID"][df.BasketID==elem]=str(new_customer_per_basket_list[i])+"N"

In [34]:
df[df["CustomerID"].isnull()]

Empty DataFrame
Columns: [BasketID, BasketDate, Sale, CustomerID, CustomerCountry, ProdID, ProdDescr, Qta]
Index: []

In [37]:
df[df["CustomerID"].astype(str).str.contains('N')]

BasketID           BasketDate     Sale CustomerID CustomerCountry  \
1444     536544  2010-12-01 14:32:00     2.51     63964N  United Kingdom   
1446     536544  2010-12-01 14:32:00     1.66     63964N  United Kingdom   
1447     536544  2010-12-01 14:32:00     1.66     63964N  United Kingdom   
1448     536544  2010-12-01 14:32:00     2.51     63964N  United Kingdom   
1450     536544  2010-12-01 14:32:00     0.43     63964N  United Kingdom   
...         ...                  ...      ...        ...             ...   
541532   581498  2011-12-09 10:26:00     4.13     55002N  United Kingdom   
541534   581498  2011-12-09 10:26:00     3.29     55002N  United Kingdom   
541538   581498  2011-12-09 10:26:00     4.96     55002N  United Kingdom   
541539   581498  2011-12-09 10:26:00    10.79     55002N  United Kingdom   
541540   581498  2011-12-09 10:26:00  1714.17     55002N  United Kingdom   

        ProdID                           ProdDescr  Qta  
1444     21774     DECORATIVE CATS BATHROOM BOTTLE    2  
1446     21787               RAIN PONCHO RETROSPOT    2  
1447     21790                  VINTAGE SNAP CARDS    9  
1448     21791  VINTAGE HEADS AND TAILS CARD GAME     2  
1450     21802     CHRISTMAS TREE HEART DECORATION    9  
...        ...                                 ...  ...  
541532   85038     6 CHOCOLATE LOVE HEART T-LIGHTS    1  
541534  85049a       TRADITIONAL CHRISTMAS RIBBONS    5  
541538   85150       LADIES & GENTLEMEN METAL SIGN    1  
541539   85174                   S/4 CACTI CANDLES    1  
541540     DOT                      DOTCOM POSTAGE    1  

[65080 rows x 8 columns]

Sono stati generati nuovi CustomerId per non buttare transazioni mancanti

## Quantità negative e non cancellati
Abbiamo già constatato che gli ordini con la C sono cancellati e abbiamo già deciso di cancellarli. Consideriamo ora quelli con solo le quantità negative. 

In [47]:
print(len(df[df["Qta"]<=0]))
print(len(df[(df["Qta"]<=0) & (df["BasketID"].str.contains("C"))]))

print(len(df[(df["Qta"]<=0) & (df["BasketID"].str.contains("C")==False)]))
print(len(df[(df["Qta"]<=0) & (df["Sale"]<=0)&(df["BasketID"].str.contains("C")==False)]))
print(df[(df["Qta"]<=0) & (df["BasketID"].str.contains("C")==False)].ProdDescr.unique())

9752
9084
668
668
[nan 'check' 'Dotcom sales' 'reverse 21/5/10 adjustment'
 'mouldy, thrown away.' '?' 'label mix up' 'samples/damages' 'thrown away'
 'damages' 'showroom' 'wrongly sold as sets' 'dotcom sold sets'
 'Amazon sold sets' 'wrongly sold sets' '?sold as sets?' 'damages/display'
 'damaged stock' 'damages?' 're dotcom quick fix.' 'sold in set?'
 'POSSIBLE DAMAGES OR LOST?' 'damaged' 'Damaged' 'DAMAGED' 'Missing'
 'wrong code?' 'crushed' 'damages/credits from ASOS.'
 'Not rcvd in 10/11/2010 delivery' 'Thrown away-rusty' 'damages/dotcom?'
 'smashed' 'reverse previous adjustment'
 'incorrectly credited C550456 see 47' 'wet damaged' 'Water damaged'
 'missing' 'sold as set on dotcom' 'mix up with c' 'ebay'
 'Sold as 1 on dotcom' 'taig adjust no stock' '?display?' '??'
 'OOPS ! adjustment' 'Dagamed' 'historic computer difference?....se'
 'incorrect stock entry.' 'wrongly coded-23343' 'stock check' 'WET/MOULDY'
 'Wet pallet-thrown away' 're-adjustment' 'wrongly coded 20713'
 'Unsaleab

Abbiamo solamente 668 transazioni in cui abbiamo la quantità negativa e non appartengono a ordini cancellati. Inoltre in questi 668 tutti hanno sale<=0.  Dalle descrizioni fornite la maggior parte sono quindi articoli persi oppure danneggiati o eventuali errori.

In [44]:
df[(df["Qta"]<=0) & (df["BasketID"].str.contains("C")==False)]

BasketID           BasketDate  Sale CustomerID CustomerCountry  ProdID  \
4347     536764  2010-12-02 14:42:00   0.0     22874N  United Kingdom  84952C   
7188     536996  2010-12-03 15:30:00   0.0     63823N  United Kingdom   22712   
7192     537000  2010-12-03 15:32:00   0.0     95404N  United Kingdom   21414   
7193     537001  2010-12-03 15:33:00   0.0     59702N  United Kingdom   21653   
7195     537003  2010-12-03 15:33:00   0.0     30662N  United Kingdom   85126   
...         ...                  ...   ...        ...             ...     ...   
535321   581198  2011-12-07 18:26:00   0.0     78679N  United Kingdom   22025   
535322   581199  2011-12-07 18:26:00   0.0     87097N  United Kingdom   84581   
535324   581201  2011-12-07 18:30:00   0.0     64996N  United Kingdom   22217   
535328   581205  2011-12-07 18:32:00   0.0      7847N  United Kingdom   20893   
535335   581212  2011-12-07 18:38:00   0.0     30484N  United Kingdom   22578   

       ProdDescr   Qta  
4347         NaN   -38  
7188         NaN   -20  
7192         NaN   -22  
7193         NaN    -6  
7195         NaN    -2  
...          ...   ...  
535321     check   -30  
535322       NaN    -2  
535324  damages?  -155  
535328   damages   -55  
535335      lost -1050  

[668 rows x 8 columns]

Verifichiamo però che non ci sia una trasazione speculare ma positiva. Proviamo con alcuni prodID

In [50]:
df[df["ProdID"]=="21414"]

BasketID           BasketDate  Sale CustomerID CustomerCountry ProdID  \
7192     537000  2010-12-03 15:32:00   0.0     95404N  United Kingdom  21414   
98375    544677  2011-02-22 16:13:00   2.1      18196  United Kingdom  21414   
147790   549136  2011-04-06 15:23:00   0.0     53902N  United Kingdom  21414   

                      ProdDescr  Qta  
7192                        NaN  -22  
98375   SCALLOP SHELL SOAP DISH    1  
147790                      NaN   -3

Non si rileva nessun particolare pattern che fa presumere la presenza di un corrispettivo ordine positivo.

*Decisione finale*: Eliminare anche questi sample a meno che non ci vien ein mente qualche indicatore che possa tenere conto di questi articoli. In ogni caso sono solo 668 elementi.